staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval_test.go

```diff
@@ -0,0 +1,486 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cacher
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+	"testing"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+func intervalFromEvents(events []*watchCacheEvent) *watchCacheInterval {
+	startIndex, endIndex, locker := 0, len(events), &sync.Mutex{}
+	indexer := func(i int) *watchCacheEvent {
+		if len(events) == 0 {
+			return nil
+		}
+		return events[i]
+	}
+	indexValidator := func(_ int) bool { return true }
+
+	return newCacheInterval(startIndex, endIndex, indexer, indexValidator, locker)
+}
+
+func bufferFromEvents(events []*watchCacheEvent) *watchCacheIntervalBuffer {
+	wcib := &watchCacheIntervalBuffer{
+		buffer:     make([]*watchCacheEvent, bufferSize),
+		startIndex: 0,
+		endIndex:   len(events),
+	}
+	copy(wcib.buffer, events)
+
+	return wcib
+}
+
+func generateEvents(start, end int) []*watchCacheEvent {
+	n := end - start
+	events := make([]*watchCacheEvent, n)
+	for i := 0; i < n; i++ {
+		events[i] = &watchCacheEvent{
+			Type: watch.Added,
+			Object: makeTestPod(
+				fmt.Sprintf("pod%d", start+i+1),
+				uint64(start+i+1),
+			),
+		}
+	}
+	return events
+}
+
+// eventVerifier consolidates logic required to verify events
+// and other return values that are returned from either the
+// watchCacheIntervalBuffer::next() or watchCacheInterval::Next()
+// APIs so that it can be reused in tests.
+type eventVerifier struct {
```

Comment:
I think it's significantly overengineered. Let's simplify this.

I would just simply make it similar to verifyNoEvent:

```
func verifyEvent(ok bool, event, expectedEvent *watchCacheEvent) error {
  if !ev.ok {
    return fmt.Errorf("expected event: %#v, got no event", expectedEvent)
  }

  /*  See my comment below why I don't think more extensive check isn't needed */
  if event == nil {
    return fmt.Errorf("unexpected nil event, expected: %#v", expectedEvent)
  }

  if !reflect.DeepEual(event, expectedEvent) {
    return fmt.Errorf("expected %v, got %v", *event, *expectedEvent)
  }

  return nil
}